suite/tests/client-interface/avx512ctx.dll.c

```diff
@@ -35,13 +35,123 @@
 #include "avx512ctx-shared.h"
 #include <string.h>
 
-#define CHECK(x, msg)                                                                \
-    do {                                                                             \
-        if (!(x)) {                                                                  \
-            dr_fprintf(STDERR, "CHECK failed %s:%d: %s\n", __FILE__, __LINE__, msg); \
-            dr_abort();                                                              \
-        }                                                                            \
-    } while (0);
+static void
+print_zmm(byte zmm_buf[], byte zmm_ref[])
+{
+    int zmm_off = 0;
+    for (int i = 0; i < NUM_SIMD_REGS; ++i, zmm_off += 64) {
+        dr_fprintf(STDERR, "got zmm[%d]:\n", i);
+        for (int b = zmm_off; b < zmm_off + 64; ++b)
+            dr_fprintf(STDERR, "%x", zmm_buf[b]);
+        dr_fprintf(STDERR, "\nref zmm[%d]:\n", i);
+        for (int b = zmm_off; b < zmm_off + 64; ++b)
+            dr_fprintf(STDERR, "%x", zmm_ref[b]);
+        dr_fprintf(STDERR, "\n");
+    }
+}
+
+static void
+print_opmask(byte opmask_buf[], byte opmask_ref[])
+{
+    int opmask_off = 0;
+    for (int i = 0; i < NUM_OPMASK_REGS; ++i, opmask_off += 8) {
+        dr_fprintf(STDERR, "got k[%i]:\n", i);
+        for (int b = opmask_off; b < opmask_off + 8; ++b)
+            dr_fprintf(STDERR, "%x", opmask_buf[b]);
+        dr_fprintf(STDERR, "\nref k[%d]:\n", i);
+        for (int b = opmask_off; b < opmask_off + 8; ++b)
+            dr_fprintf(STDERR, "%x", opmask_ref[b]);
+        dr_fprintf(STDERR, "\n");
+    }
+}
+
+static void
+read_avx512_state()
+{
+    byte zmm_buf[NUM_SIMD_REGS * 64];
+    byte zmm_ref[NUM_SIMD_REGS * 64];
+    byte opmask_buf[NUM_OPMASK_REGS * 8];
+    byte opmask_ref[NUM_OPMASK_REGS * 8];
+    memset(zmm_buf, 0xde, sizeof(zmm_buf));
+    memset(zmm_ref, 0xab, sizeof(zmm_ref));
+    memset(opmask_buf, 0xde, sizeof(opmask_buf));
+    memset(opmask_ref, 0, sizeof(opmask_ref));
+    for (int i = 0; i < NUM_OPMASK_REGS; ++i) {
+        /* See comment in applicaton part of the test: We limit the test
+         * to 2 byte wide mask registers.
+         */
+        memset(&opmask_ref[i * 8], 0xabab, 2);
+    }
+
+    dr_fprintf(STDERR, "Reading application state\n");
+
+    void *drcontext = dr_get_current_drcontext();
+    dr_mcontext_t mcontext = {
+        sizeof(mcontext),
+        DR_MC_ALL,
+    };
+    dr_get_mcontext(drcontext, &mcontext);
+
+    bool get_reg_value_ok = true;
+    get_reg_value_ok &= reg_get_value_ex(DR_REG_ZMM0, &mcontext, &zmm_buf[0 * 64]);
```

Comment:
bug: in C bool is just an int and bitwise operators are problematic.  reg_get_value_ex could return 0x2 and we'd end up with 0==false.  Use logical `&&` not bitwise.

There was a similar issue in a recent PR as well.  Please keep in mind.

Reply:
Thanks. Ideally since we _defined_ bool I thought the contract with a function that returns bool is to return only true or false as defined by us, i.e. 0 or 1. I see that is hard to guarantee w/o the language behind it. I'll fix. Worth adding to https://github.com/DynamoRIO/dynamorio/wiki/Code-Style. => no bit operators on bool?

Reply:
Fixed.

Reply:
Not sure it's at the level of a code style convention when it's a bug to do otherwise so it's not really a choice.